### PR TITLE
fix(auth): Allow retries with verifyTotpSetup()

### PR DESCRIPTION
**Pull request by Equartey**
*Issue #, if available:*
https://github.com/aws-amplify/amplify-flutter/issues/4490

*Description of changes:*
Catches and correctly handles missed `EnableSoftwareTokenMFAException` when entering a wrong code during TOTP setup. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart [hunk 1]
```diff
@@ -59,29 +62,52 @@ final class TotpSetupStateMachine
...
+        emit(
+          TotpSetupState.requiresVerification(
+            _details!,
           ),
-        )
-        .result;
+        );
+        return;
+      }
+      logger.error(
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:103]: We are catching and swallowing all exceptions other than `EnableSoftwareTokenMfaException`, correct? Is that what we want to do? I was expecting the exception to be rethrown.

Comment (Equartey) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:110]: Good point. I made the exception set the state to failure which should bubble up the concern.

**Review by Jordan-Nelson — DISMISSED** [review]

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:1]: do we need to do similar thing in `sign_in_state_machine.dart` in `Future<void> verifySoftwareToken` method?

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart [hunk 2]
```diff
@@ -59,29 +62,59 @@ final class TotpSetupStateMachine
...
+        'Failed to verify TOTP code. Please try again. If the problem persists, contact support.',
```
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:111]: not sure if we should include `If the problem persists, contact support.` in the error message

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart [hunk 3]
```diff
@@ -59,29 +62,59 @@ final class TotpSetupStateMachine
...
+              session: _session,
+              userCode: event.code,
+              friendlyDeviceName: event.friendlyDeviceName,
+            ),
+          )
+          .result;
+    } on Exception catch (e, st) {
+      // Handle mismatch code exception that may occur during TOTP verification.
+      // See: https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_VerifySoftwareToken.html#API_VerifySoftwareToken_Errors
+      // TODO(equartey): Investigate when CodeMismatchException is thrown and handle accordingly.
```
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:93]: reading the exceptions' discription seems the `CodeMismatchException` is thrown when the totp is already setup and later when signing up the user using a wrong code. I'm not very familiar with our auth state machine so not sure if we have different flow/state machine for setting up the totp and then using totp code in the sign-in flow

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart
Comment (Equartey) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:1]: I don't believe this is a concern. The sign in flow currently allows multiple attempts. That being said, looking at that function I don't know offhand what catches the `EnableSoftwareTokenMfaException` that enables the retries.

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart (hunk 3 above)
Comment (Equartey) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart:93]: I was curious about this too, though I could not get that exception to throw. Our state machine blocks calls to `Amplify.Auth.verifyTotpSetup()` when the client is not in the setup process. After an authenticator app is setup, the totp code gets submitted during sign. As it's setup now, I don't think this is a concern.

Comment (Equartey) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart]: note: I omitted the TODO comment after reviewing it more.

**Review by NikaHsn — APPROVED** [review]

**Review by khatruong2009 — APPROVED** [review]

**Pull request merged**